### PR TITLE
RAML 1.0: Support for multiple examples

### PR DESCRIPTION
**Pull request by juancoen**
screen shot 2016-03-11 at 5 48 23 pm
screen shot 2016-03-11 at 5 48 28 pm

### CODE REVIEW

File: src/app/directives/documentation.tpl.html [hunk 1]
```diff
@@ -186,6 +186,33 @@ <h4 class="raml-console-resource-body-heading">
             <pre class="raml-console-resource-pre"><code class="raml-console-hljs" hljs source="getBeatifiedExample(responseInfo[code][responseInfo[code].currentType].example)"></code></pre>
           </div>
 
+          <div ng-if="responseInfo[code][responseInfo[code].currentType].examples">
+            <span>Examples:</span>
+            <div ng-if="isArray(responseInfo[code][responseInfo[code].currentType].examples)">
```
Comment (xaka) [inline src/app/directives/documentation.tpl.html]: Let's convert it to an array at controller level and assume we always deal with array at template level. It's going to simplify it a lot.

Comment (juancoen) [inline src/app/directives/documentation.tpl.html]: We can do that, but we will lose the example name. For example, with this RAML:
```
/resource:
  get:
    responses:
      200:
        body:
          application/json:
            examples:
              one:
                content: |
                  {
                    "property1": "value1"
                  }
              two:
                content: |
                  {
                    "property2": "value2"
                  }
```

the use of 'one' and 'two' to distinguish the examples will be lost in the console.

Comment (xaka) [inline src/app/directives/documentation.tpl.html]: In this case it's somebody else responsibility to make sure data is consumable by template. Let's say there is requirement (contract) from template - it expects to receive an array of objects where each object has "name" and "content" property. Now we need to deal with following cases:

- when there is only "example" property, which is string, we need to convert it to array of one element with generated name
- when there is "examples" as array, we need to generate names
- when there is "examples" as object, we need to sort names and generate array

Comment (juancoen) [inline src/app/directives/documentation.tpl.html]: :+1:

File: src/app/directives/documentation.js [hunk 2]
```diff
@@ -169,6 +169,40 @@
...
+            if (Array.isArray(selectedResponse.examples)) {
+              return selectedResponse.examples.map(function (example, index) {
+                console.log(index);
+                return {
+                  name: 'Example ' + index,
+                  content: example
+                };
+              });
+            } else {
+              return Object.keys(selectedResponse.examples).map(function (key) {
```
Comment (xaka) [inline src/app/directives/documentation.js]: Don't forge to sort to get predicable order.

File: src/app/directives/documentation.js [hunk 3]
```diff
@@ -169,6 +169,40 @@
...
+        function transformExample(selectedResponse) {
+          if (selectedResponse.example) {
+            return [{
+              name: 'Example',
+              content: selectedResponse.example
+            }];
+          } else if (selectedResponse.examples) {
+            if (Array.isArray(selectedResponse.examples)) {
+              return selectedResponse.examples.map(function (example, index) {
+                console.log(index);
```
Comment (xaka) [inline src/app/directives/documentation.js]: console.log detected

File: src/app/directives/documentation.tpl.html [hunk 4]
```diff
@@ -181,9 +181,17 @@ <h4 class="raml-console-resource-body-heading">
...
           </h4>
 
-          <div ng-if="responseInfo[code][responseInfo[code].currentType].example">
-            <span>Example:</span>
-            <pre class="raml-console-resource-pre"><code class="raml-console-hljs" hljs source="getBeatifiedExample(responseInfo[code][responseInfo[code].currentType].example)"></code></pre>
+          <div ng-if="examples">
+            <span>Examples:</span>
+            <span
+              class="raml-console-flag"
+              ng-class="{ 'raml-console-is-active': currentExample === $index }"
```
Comment (xaka) [inline src/app/directives/documentation.tpl.html]: Did you mean `is-active`? without `raml-console` prefix as in SCSS you have:
```
&.is-active {
  ...
}
```

File: src/app/directives/documentation.tpl.html [hunk 5]
```diff
@@ -181,9 +181,17 @@ <h4 class="raml-console-resource-body-heading">
...
+              ng-repeat="example in examples"
+              ng-click="changeExample($index)">{{example.name}}</span>
+
+            <pre ng-if="currentExample === $index" class="raml-console-resource-pre" ng-repeat="example in examples">
```
Comment (xaka) [inline src/app/directives/documentation.tpl.html]: I don't think you need `ng-repeat` and `ng-if` here. Just use `examples[currentExample]`.

File: src/app/directives/documentation.tpl.html (hunk 4 above)
Comment (juancoen) [inline src/app/directives/documentation.tpl.html]: the raml-console prefix is added as part of scss -> css transformation

File: src/app/directives/documentation.tpl.html [hunk 6]
```diff
@@ -181,9 +181,17 @@ <h4 class="raml-console-resource-body-heading">
             <span ng-click="changeType($event, key, code)" ng-class="{ 'raml-console-is-active': $first}" class="raml-console-flag" ng-repeat="(key, value) in methodInfo.responses[code].body">{{key}}</span>
           </h4>
 
-          <div ng-if="responseInfo[code][responseInfo[code].currentType].example">
-            <span>Example:</span>
-            <pre class="raml-console-resource-pre"><code class="raml-console-hljs" hljs source="getBeatifiedExample(responseInfo[code][responseInfo[code].currentType].example)"></code></pre>
+          <div ng-if="examples">
```
Comment (xaka) [inline src/app/directives/documentation.tpl.html]: You need to be explicit and check `.length` property

File: src/app/directives/documentation.tpl.html (hunk 4 above)
Comment (xaka) [inline src/app/directives/documentation.tpl.html]: OMG. Alright.

File: src/app/directives/documentation.js [hunk 7]
```diff
@@ -169,6 +169,39 @@
...
+            } else {
+              return Object.keys(selectedResponse.examples).sort().map(function (key) {
+                return {
+                  name: key,
+                  content: selectedResponse.examples[key]
+                };
+              });
+            }
+          }
+        }
```
Comment (xaka) [inline src/app/directives/documentation.js]: Unit test please :)

File: src/app/directives/documentation.tpl.html (hunk 6 above)
Comment (juancoen) [inline src/app/directives/documentation.tpl.html]: I don't think so, if example and examples are undefined, $scope.examples will be also undefined. I can change the transformExample to return an empty array when example and examples are undefined

Comment (xaka) [inline src/app/directives/documentation.tpl.html]: It's ok for `ng-if`, it handles undefined as falsy.

Comment (juancoen) [inline src/app/directives/documentation.tpl.html]: ok, so I'm changing the ng-if check to look like: `examples && examples.length`

**Comment by xaka** [comment]
LGTM :+1:

**Pull request merged**